app/src/main/java/com/odysee/app/ui/findcontent/FileViewFragment.java

```diff
@@ -1901,14 +1940,71 @@ public void onPlayWhenReadyChanged(boolean playWhenReady, int reason) {
                             userNotStreaming.setVisibility(View.VISIBLE);
                         }
                     }
+
+                    if (mediaSource != null) {
+                        MainActivity.appPlayer.setMediaSource(mediaSource, true);
+                        MainActivity.appPlayer.prepare();
+                    }
                 }
+            }
+        }
+    }
 
-                if (mediaSource != null) {
-                    MainActivity.appPlayer.setMediaSource(mediaSource, true);
-                    MainActivity.appPlayer.prepare();
+    private void getStreamingUrlAndInitializePlayer() {
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        executor.execute(() -> {
+            try {
+                // Get the streaming URL
+                Map<String, Object> params = new HashMap<>();
+                params.put("uri", claim.getPermanentUrl());
+                JSONObject result = (JSONObject) Lbry.parseResponse(Lbry.apiCall(Lbry.METHOD_GET, params));
+                String sourceUrl = (String) result.get("streaming_url");
+
+                // Get the stream type
+                OkHttpClient client = new OkHttpClient();
+                Request request = new Request.Builder()
+                        .url(sourceUrl)
+                        .head()
+                        .build();
+                try (Response response = client.newCall(request).execute()) {
+                    String contentType = response.header("Content-Type");
+                    MainActivity.videoIsTranscoded = contentType.equals("application/x-mpegurl"); // m3u8
                 }
+
+                new Handler(Looper.getMainLooper()).post(() -> initializePlayer(sourceUrl));
+            } catch (LbryRequestException | LbryResponseException | JSONException | IOException ex) {
+                // TODO: How does error handling work here
+                ex.printStackTrace();
             }
+        });
+        executor.shutdown();
```

Comment:
Like this?

Reply:
There are different ways to do background work on Java. I normally create a thread and I use the Future API to run the code. That allows me to shutdown the executor when the task has finished, as the background thread will be blocked while the UI thread will not be.

You could left it as it is now, if it is not crashing the app or try to refactor the code to use a Future.

shutdown() wait for tasks which were started already -submitted AND started-; if the task was not started, your code will not run as the executor will be shutdown.

Executors are kept on memory. That's why they must be shutdown. Odysee have had them without shutting them down and it worked fine. This is more a performance improvement.

Reply:
It seems to be working fine, so I'll just leave it as it is.